src/cholmod.jl

```diff
@@ -88,144 +93,9 @@ struct Method
-    metis_memory::Cdouble
-    metis_dswitch::Cdouble
-    metis_nswitch::Csize_t
-    nrow::Csize_t
-    mark::SuiteSparse_long
-    iworksize::Csize_t
-    xworksize::Csize_t
-    Flag::Ptr{Nothing}
-    Head::Ptr{Nothing}
-    Xwork::Ptr{Nothing}
-    Iwork::Ptr{Nothing}
-    itype::Cint
-    dtype::Cint
-    no_workspace_reallocate::Cint
-    status::Cint
-    fl::Cdouble
-    lnz::Cdouble
-    anz::Cdouble
-    modfl::Cdouble
-    malloc_count::Csize_t
-    memory_usage::Csize_t
-    memory_inuse::Csize_t
-    nrealloc_col::Cdouble
-    nrealloc_factor::Cdouble
-    ndbounds_hit::Cdouble
-    rowfacfl::Cdouble
-    aatfl::Cdouble
-    called_nd::Cint
-    blas_ok::Cint
-    SPQR_grain::Cdouble
-    SPQR_small::Cdouble
-    SPQR_shrink::Cint
-    SPQR_nthreads::Cint
-    SPQR_flopcount::Cdouble
-    SPQR_analyze_time::Cdouble
-    SPQR_factorize_time::Cdouble
-    SPQR_solve_time::Cdouble
-    SPQR_flopcount_bound::Cdouble
-    SPQR_tol_used::Cdouble
-    SPQR_norm_E_fro::Cdouble
-    SPQR_istat::NTuple{10,SuiteSparse_long}
-    useGPU::Cint
-    maxGpuMemBytes::Csize_t
-    maxGpuMemFraction::Cdouble
-    gpuMemorySize::Csize_t
-    gpuKernelTime::Cdouble
-    gpuFlops::SuiteSparse_long
-    gpuNumKernelLaunches::Cint
-    cublasHandle::Ptr{Nothing}
-    gpuStream::NTuple{8,Ptr{Nothing}}
-    cublasEventPotrf::NTuple{3,Ptr{Nothing}}
-    updateCKernelsComplete::Ptr{Nothing}
-    updateCBuffersFree::NTuple{8,Ptr{Nothing}}
-    dev_mempool::Ptr{Nothing}
-    dev_mempool_size::Csize_t
-    host_pinned_mempool::Ptr{Nothing}
-    host_pinned_mempool_size::Csize_t
-    devBuffSize::Csize_t
-    ibuffer::Cint
-    syrkStart::Cdouble
-    cholmod_cpu_gemm_time::Cdouble
-    cholmod_cpu_syrk_time::Cdouble
-    cholmod_cpu_trsm_time::Cdouble
-    cholmod_cpu_potrf_time::Cdouble
-    cholmod_gpu_gemm_time::Cdouble
-    cholmod_gpu_syrk_time::Cdouble
-    cholmod_gpu_trsm_time::Cdouble
-    cholmod_gpu_potrf_time::Cdouble
-    cholmod_assemble_time::Cdouble
-    cholmod_assemble_time2::Cdouble
-    cholmod_cpu_gemm_calls::Csize_t
-    cholmod_cpu_syrk_calls::Csize_t
-    cholmod_cpu_trsm_calls::Csize_t
-    cholmod_cpu_potrf_calls::Csize_t
-    cholmod_gpu_gemm_calls::Csize_t
-    cholmod_gpu_syrk_calls::Csize_t
-    cholmod_gpu_trsm_calls::Csize_t
-    cholmod_gpu_potrf_calls::Csize_t
-
-    function Common()
-        common = new()
-        ccall((@cholmod_name("start"), :libcholmod),
-            Cint, (Ptr{Common},), common)
-        finalizer(common) do common
-            ccall((@cholmod_name("finish"), :libcholmod),
-                  Cint, (Ptr{Common},), common) == TRUE
-        end
-    end
-end
-
-Base.unsafe_convert(::Type{Ptr{Common}},c::Common) =
-    reinterpret(Ptr{Common}, pointer_from_objref(c))
-
-const common = Vector{Common}()
-
-function defaults!(common)
-    ccall((@cholmod_name("defaults"), :libcholmod),
-        Cint, (Ptr{Common},), common)
-    return common
-end
+const COMMONS = Vector{Ref{cholmod_common}}()
```

Comment:
Why the extra `Ref` here?

Reply:
Explicitly show we're using a ref `&` instead of implicitly do the conversion using `pointer_from_objref` in `Base.unsafe_convert`.

Reply:
In the old style, the `ccall` expects `Ptr{cholmod_common}`, but we just pass a `cholmod_common` to it, which feels like a little bit wired.

Reply:
Mh, okay. Is this the standard way to do this sort of thing? I'm not convinced as yet, for the following reasons:
 - Julia clearly owns the `COMMONS` object, and calling a SuiteSparse C function `foo(common)` should be roughly the same as calling an equivalent `foo!(common)` Julia function. You wouldn't use a `Ref` for this sort of thing. 
 - Adding the `[]` after every reference to a `COMMONS[k]` adds extra noise and can be frustrating if you forget. 
Happy to let this one go if your convinced of your solution.

Reply:
Hi @yuyichao, could you chime in and give some caveats? 

The implicit way mentioned above is to add the following method to `Base.unsafe_convert`.
```
Base.unsafe_convert(::Type{Ptr{Common}},c::Common) = reinterpret(Ptr{Common}, pointer_from_objref(c))
```